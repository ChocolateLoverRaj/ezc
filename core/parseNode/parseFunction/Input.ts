import EnumItemWithData from '../../EnumItemWithData'
import TryParseNode from '../TryParseNode'
import KeyWordsToInputFlags from '../parseInputFlags/Input'

interface Input {
  typeParsers: ReadonlyArray<TryParseNode<EnumItemWithData>>
  keyWordsToInputFlags: KeyWordsToInputFlags
  parseBlock: TryParseNode<EnumItemWithData>
}

export default Input
