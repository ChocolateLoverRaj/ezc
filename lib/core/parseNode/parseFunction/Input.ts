import EnumItemWithData from '../../EnumItemWithData'
import TryParseNode from '../TryParseNode'
import KeyWordsToInputFlags from '../parseInputFlags/Input'

interface Input {
  typeParsers: ReadonlyArray<TryParseNode<EnumItemWithData>>
  keyWordsToInputFlags: KeyWordsToInputFlags
}

export default Input
