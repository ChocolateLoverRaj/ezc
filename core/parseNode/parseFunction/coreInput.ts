import coreTypeParsers from '../coreTypeParsers'
import coreInstructionParsers from '../parseBlock/coreInstructionParsers'
import parseBlock from '../parseBlock/parseBlock'
import coreKeyWordsToInputFlags from '../parseInputFlags/coreKeyWordsToInputFlags'
import Input from './Input'

const coreInput: Input = {
  typeParsers: coreTypeParsers,
  keyWordsToInputFlags: coreKeyWordsToInputFlags,
  parseBlock: parseBlock(coreInstructionParsers)
}

export default coreInput
