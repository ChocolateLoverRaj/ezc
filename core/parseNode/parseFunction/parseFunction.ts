import CoreTokensWithData from '../../CoreTokensWithData'
import CoreTokenType from '../../CoreTokenType'
import CoreKeyWord from '../../CoreKeyWord'
import CoreNodesWithData from '../CoreNodesWithData'
import CoreNodeType from '../CoreNodeType'
import TryParseNode from '../TryParseNode'
import Input from './Input'
import splitAsyncIterator from '../../../util/splitAsyncIterator/splitAsyncIterator'
import skipSplittedIterator from '../../../util/splitAsyncIterator/skip'
import tryNodeParsers from '../tryNodeParsers/tryNodeParsers'
import parseIdentifier from '../parseIdentifier'
import EnumItemWithData from '../../EnumItemWithData'
import parseInput from '../parseInput/parseInput'
import checkKeyWord from '../checkKeyWord'

const parseFunction = (
  { typeParsers, keyWordsToInputFlags, parseBlock }: Input
): TryParseNode<CoreNodesWithData[CoreNodeType.FUNCTION]> => async stream => {
  const splittedIterator = splitAsyncIterator(stream[Symbol.asyncIterator]())
  const type = {
    enum: CoreNodeType,
    id: CoreNodeType.FUNCTION
  }

  let length = 0
  const skip = (numberOfElements: number): void => {
    skipSplittedIterator(splittedIterator, numberOfElements)
    length += numberOfElements
  }

  {
    const error = await checkKeyWord(
      splittedIterator.asyncIterable[Symbol.asyncIterator](),
      type,
      length,
      'Expected define',
      { enum: CoreKeyWord, id: CoreKeyWord.DEFINE }
    )
    if (error !== undefined) return error
    skip(1)
  }

  const parseReturnTypeResult = await tryNodeParsers(typeParsers)(splittedIterator.asyncIterable)
  if (!parseReturnTypeResult.success) {
    return {
      success: false,
      result: {
        type,
        index: length,
        message: 'Expected return type',
        subAttempts: parseReturnTypeResult.result
      }
    }
  }
  skip(parseReturnTypeResult.result.length)

  const parseIdentifierResult = await parseIdentifier(splittedIterator.asyncIterable)
  if (!parseIdentifierResult.success) {
    return {
      success: false,
      result: {
        type,
        index: length,
        message: 'Expected identifier',
        subAttempts: [parseIdentifierResult.result]
      }
    }
  }
  skip(parseIdentifierResult.result.length)

  // Parse (
  {
    const error = await checkKeyWord(
      splittedIterator.asyncIterable[Symbol.asyncIterator](),
      type,
      length,
      'Expected open parenthesis',
      { enum: CoreKeyWord, id: CoreKeyWord.OPEN_PARENTHESIS }
    )
    if (error !== undefined) return error
    skip(1)
  }

  // Parse inputs
  const inputs: EnumItemWithData[] = []
  for (let firstInput = true; ; firstInput = false) {
    // Check for )
    {
      const { done, value } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
      if (done === true) {
        return {
          success: false,
          result: {
            type,
            index: length,
            message: 'Expected closing parenthesis',
            subAttempts: undefined
          }
        }
      }
      if (value.type.enum === CoreTokenType && value.type.id === CoreTokenType.KEY_WORD) {
        const { data } = value as CoreTokensWithData[CoreTokenType.KEY_WORD]
        if (data.enum === CoreKeyWord && data.id === CoreKeyWord.CLOSE_PARENTHESIS) {
          skip(1)
          break
        }
      }
    }

    // Check for ,
    if (!firstInput) {
      const error = await checkKeyWord(
        splittedIterator.asyncIterable[Symbol.asyncIterator](),
        type,
        length,
        'Expected comma after input',
        { enum: CoreKeyWord, id: CoreKeyWord.COMMA }
      )
      if (error !== undefined) return error
      skip(1)
    }

    const parseInputInput = { typeParsers, keyWordsToInputFlags }
    const parseInputResult = await parseInput(parseInputInput)(splittedIterator.asyncIterable)
    if (!parseInputResult.success) {
      return {
        success: false,
        result: {
          type,
          index: length,
          message: 'Expected input',
          subAttempts: [parseInputResult.result]
        }
      }
    }
    skip(parseInputResult.result.length)
    inputs.push(parseInputResult.result.node)
  }

  // Parse {
  {
    const error = await checkKeyWord(
      splittedIterator.asyncIterable[Symbol.asyncIterator](),
      type,
      length,
      'Expected open curly bracket',
      { enum: CoreKeyWord, id: CoreKeyWord.OPEN_CURLY_BRACKET }
    )
    if (error !== undefined) return error
    skip(1)
  }

  // Parse blocks
  const blocks: EnumItemWithData[] = []
  const parseBlockError = await (async () => {
    while (true) {
      const parseBlockResult = await parseBlock(splittedIterator.asyncIterable)
      if (!parseBlockResult.success) return parseBlockResult.result
      skip(parseBlockResult.result.length)
      blocks.push(parseBlockResult.result.node)
    }
  })()

  // Parse }
  {
    const error = await checkKeyWord(
      splittedIterator.asyncIterable[Symbol.asyncIterator](),
      type,
      length,
      'Expected close curly bracket',
      { enum: CoreKeyWord, id: CoreKeyWord.CLOSE_CURLY_BRACKET }
    )
    if (error !== undefined) {
      return {
        success: false,
        result: {
          type,
          index: length,
          message: 'Error parsing blocks',
          subAttempts: [parseBlockError, error.result]
        }
      }
    }
    skip(1)
  }

  return {
    success: true,
    result: {
      node: {
        type,
        data: {
          name: parseIdentifierResult.result.node,
          inputs,
          returnType: parseReturnTypeResult.result.node,
          blocks
        }
      },
      length
    }
  }
}

export default parseFunction
