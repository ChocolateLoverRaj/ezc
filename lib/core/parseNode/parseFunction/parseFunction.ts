import CoreTokensWithData from '../../CoreTokensWithData'
import CoreTokenType from '../../CoreTokenType'
import CoreKeyWord from '../../CoreKeyWord'
import CoreNodesWithData from '../CoreNodesWithData'
import CoreNodeType from '../CoreNodeType'
import TryParseNode from '../TryParseNode'
import Input from './Input'
import splitAsyncIterator from '../../splitAsyncIterator/splitAsyncIterator'
import skipSplittedIterator from '../../splitAsyncIterator/skip'
import tryNodeParsers from '../tryNodeParsers'
import parseIdentifier from '../parseIdentifier'
import CoreTokenDatas from '../../CoreTokenDatas'
import OpenCloseType from '../../OpenCloseType'
import InputType from '../InputType'
import parseInputFlags from '../parseInputFlags/parseInputFlags'

const parseFunction = (
  { typeParsers, keyWordsToInputFlags }: Input
): TryParseNode<CoreNodesWithData[CoreNodeType.FUNCTION]> => async stream => {
  const splittedIterator = splitAsyncIterator(stream[Symbol.asyncIterator]())

  let length = 0
  const skip = (numberOfElements: number): void => {
    skipSplittedIterator(splittedIterator, numberOfElements)
    length += numberOfElements
  }

  {
    const { done, value } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
    if (done === true) return
    if (!(value.type.enum === CoreTokenType && value.type.id === CoreTokenType.KEY_WORD)) return
    const { data } = value as CoreTokensWithData[CoreTokenType.KEY_WORD]
    if (!(data.enum === CoreKeyWord && data.id === CoreKeyWord.DEFINE)) return
    skip(1)
  }

  const parsedReturnType = await tryNodeParsers(typeParsers)(splittedIterator.asyncIterable)
  if (parsedReturnType === undefined) return
  skip(parsedReturnType.length)

  const parsedIdentifier = await parseIdentifier(splittedIterator.asyncIterable)
  if (parsedIdentifier === undefined) return
  skip(parsedIdentifier.length)

  {
    const { done, value } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
    if (done === true) return
    if (!(value.type.enum === CoreTokenType && value.type.id === CoreTokenType.OPEN_CLOSE)) return
    const { type, close } = value.data as CoreTokenDatas[CoreTokenType.OPEN_CLOSE]
    if (!(type === OpenCloseType.PARENTHESIS && !close)) return
    skip(1)
  }

  // Parse inputs
  const inputTypes: InputType[] = []
  const inputNames: string[] = []
  while (true) {
    // Check for )
    {
      const { done, value } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
      if (done === true) return
      if (value.type.enum === CoreTokenType && value.type.id === CoreTokenType.OPEN_CLOSE) {
        const { type, close } = value.data as CoreTokenDatas[CoreTokenType.OPEN_CLOSE]
        if (type === OpenCloseType.PARENTHESIS && close) {
          skip(1)
          break
        }
      }
    }

    const parsedType = await tryNodeParsers(typeParsers)(splittedIterator.asyncIterable)
    if (parsedType === undefined) return
    skip(parsedType.length)

    const flags = await parseInputFlags(keyWordsToInputFlags)(splittedIterator.asyncIterable)
    skip(flags.length)

    inputTypes.push({
      flags: flags.flags,
      type: parsedType.node
    })

    const parsedIdentifier = await parseIdentifier(splittedIterator.asyncIterable)
    if (parsedIdentifier === undefined) return
    skip(parsedIdentifier.length)

    // FIXME: @identifiers should be allowed, but in this code they are
    inputNames.push(parsedIdentifier.node.data.name)

    {
      const { done, value } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
      if (
        done !== true &&
        value.type.enum === CoreTokenType &&
        value.type.id === CoreTokenType.KEY_WORD
      ) {
        const { data } = value as CoreTokensWithData[CoreTokenType.KEY_WORD]
        if (data.enum === CoreKeyWord && data.id === CoreKeyWord.COMMA) {
          skip(1)
        }
      }
    }
  }

  {
    const { done, value } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
    if (done === true) return
    if (!(value.type.enum === CoreTokenType && value.type.id === CoreTokenType.OPEN_CLOSE)) return
    const { type, close } = value.data as CoreTokenDatas[CoreTokenType.OPEN_CLOSE]
    if (!(type === OpenCloseType.CURLY_BRACKET && !close)) return
    skip(1)
  }

  return {
    node: {
      type: {
        enum: CoreNodeType,
        id: CoreNodeType.FUNCTION
      },
      data: {
        name: parsedIdentifier.node.data.name,
        inputNames,
        type: {
          inputTypes,
          returnType: parsedReturnType.node
        },
        blocks: []
      }
    },
    length
  }
}

export default parseFunction
