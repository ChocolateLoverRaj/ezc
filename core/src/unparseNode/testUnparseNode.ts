import TryParseNode from '../parseNode/TryParseNode'
import EnumItemWithData from '../EnumItemWithData'
import UnparsedPartType from './UnparsedPartType'
import unparsedNodeToString from '../unparsedNodeToString/unparsedNodeToString'
import coreInput from '../unparsedNodeToString/coreInput'
import parseAllTokens from '../tryGetToken/parseAllTokens'
import coreTryers from '../tryGetToken/coreTryers'
import arrayToAsyncIterable from 'util/dist/arrayToAsyncIterable/arrayToAsyncIterable.js'

const testUnparseNode = async (
  parseNode: TryParseNode<EnumItemWithData>,
  code: string
): Promise<void> => {
  expect(unparsedNodeToString(coreInput)([{
    type: UnparsedPartType.NODE,
    data: ((await parseNode(
      parseAllTokens(coreTryers)(
        arrayToAsyncIterable([code])) as any))?.result as any).node as EnumItemWithData
  }])).toMatchSnapshot()
}

export default testUnparseNode
