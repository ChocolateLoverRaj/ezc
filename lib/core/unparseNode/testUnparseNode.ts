import TryParseNode from '../parseNode/TryParseNode'
import EnumItemWithData from '../EnumItemWithData'
import UnparsedPartType from './UnparsedPartType'
import unparsedNodeToString from '../unparsedNodeToString/unparsedNodeToString'
import coreInput from '../unparsedNodeToString/coreInput'
import parseAllTokens from '../tryGetToken/parseAllTokens'
import coreTryers from '../tryGetToken/coreTryers'
import arrayToAsyncIterable from '../arrayToAsyncIterable'

const testUnparseNode = async (
  parseNode: TryParseNode<EnumItemWithData>,
  code: string
): Promise<void> => {
  expect(unparsedNodeToString(coreInput)([{
    type: UnparsedPartType.NODE,
    data: (await parseNode(
      parseAllTokens(coreTryers)(
        arrayToAsyncIterable([code])) as any))?.node as EnumItemWithData
  }])).toMatchSnapshot()
}

export default testUnparseNode
