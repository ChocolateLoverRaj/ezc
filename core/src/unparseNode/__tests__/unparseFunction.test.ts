import parseFnCoreInput from '../../parseNode/parseFunction/coreInput'
import parseFunction from '../../parseNode/parseFunction/parseFunction'
import testUnparseNode from '../testUnparseNode'

test('simple function', async () => {
  await testUnparseNode(parseFunction(parseFnCoreInput),
    `define i1 @main() {
      0:
        ret i1 0
    }`)
})
