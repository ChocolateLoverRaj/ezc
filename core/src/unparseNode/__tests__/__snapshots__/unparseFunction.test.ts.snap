// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple function 1`] = `
"define i1 @main() {  
  0:  
    ret i1 0
}"
`;
